sonata.1
.TH SONATA 1 "October 20, 2006"
.SH NAME
sonata \- GTK+ client for the Music Player Daemon
.SH SYNOPSIS
.B sonata
[\fIOPTION\fR]
.SH DESCRIPTION
Sonata is an elegant GTK+ music client for the Music Player Daemon (MPD).
.SH OPTIONS
.TP
.B \-h\fR, \fB\-\-help
Show this help and exit
.TP
.B \-p\fR, \fB\-\-popup
Popup song notification (requires D\-Bus)
.TP
.B \-t\fR, \fB\-\-toggle
Toggles whether the app is minimized to tray or visible (requires D\-Bus)
.TP
.B \-v\fR, \fB\-\-version
Show version information and exit
.TP
.B \-\-hidden
Start app hidden (requires systray)
.TP
.B \-\-visible
Start app visible (requires systray)
.TP
.B \-\-profile\=[NUM]
Start with profile [NUM]
.TP
.B play
Play song in playlist
.TP
.B pause
Pause currently playing song
.TP
.B stop
Stop currently playing song
.TP
.B next
Play next song in playlist
.TP
.B prev
Play previous song in playlist
.TP
.B pp
Toggle play/pause; plays if stopped
.TP
.B repeat
Toggle repeat mode
.TP
.B random
Toggle random mode
.TP
.B info
Display current song info
.TP
.B status
Display MPD status
.SH "SEE ALSO"
.PP
Website: http://www.nongnu.org/sonata/
.PP
MPD Website: http://www.musicpd.org
.SH AUTHOR
Sonata was written by Scott Horowitz <stonecrest@gmail.com>.

.PP
This manual page is currently maintained by Scott Horowitz 
<stonecrest@gmail.com>. It was originally written by Michal Cihar 
<michal@cihar.com>.


CHANGELOG
Future release (not yet released!)
----------------------------------

New features
''''''''''''

    * redirect to the localized Wikipedia according to the user's local,
      instead of redirecting to the English one.
      https://github.com/multani/sonata/issues/61
    * add "Shuffle by album"
      https://github.com/multani/sonata/pull/59


Bug fixes
'''''''''

    * fix crashs on Gtk.TextView when updating the lyrics of a song
    * Fix streams parsing:
      https://github.com/multani/sonata/issues/41
      https://github.com/multani/sonata/issues/38
    * fix withdrawn mode by fixing incorrect usage of threads and Gtk
      eco-system
    * choosing a local artwork wasn't working anymore
    * fix artwork caching
    * keep the selected tab when hiding/showing Sonata using the tray icon
    * correctly handle tracks without a definitely reasonable track or disc
      value (Peter Helbing)
      https://github.com/multani/sonata/pull/74
    * fix translation of of Glade dialogs (Peter Helbing)
      https://github.com/multani/sonata/pull/75

Internal changes
''''''''''''''''

    * slightly better performance when updating the library
    * threading improvements:

      - use GIO instead of a thread to download drag-and-droped covers
      - remove the MPD from a dedicated thread. It works as good (if not
        better) as before since the threaded code was poorly written anyway and
        rendered the additionnal basically useless. This move is not definitive
        but something better will probably come up some day.

    * Looking for covers isn't as filesystem expensive as before, but it
      doesn't check for multiple possible cases in the cover name anymore.


1.7a2 (2013-11-26)
------------------

New features
''''''''''''

    * Download bigger covers from Last.fm
    * Centering the current song using Ctrl+i now also selects the current song
    * Use Transifex as a translation plateform. See
      https://www.transifex.com/projects/p/sonata/
    * Update translations:

        - Russian (thanks Stas!)
        - Ukrainian (thanks Stas!)
        - French (thanks Jon!)

Bug fixes
'''''''''

    * Fix search in library (using a specific filter or the 'everything' filter)
    * Fix scrobbling to Last.fm
    * Scrolling with the mouse wheel on the tray icon correctly changes the volume
    * Some lyrics from LyricsWiki redirects to another page. The plugin now
      follows those redirect to find the actual lyrics.
    * Fix multiple issues related to drag-and-drop of songs in the current playlist
    * Fix error while using Sonata's command line interface
    * Fix filtering songs in the current playlist
    * Fix DND of custom artwork images in the main window
    * Fix breadcrumb icon in the library while browsing an album while in Artist view
    * Fix saving the visibility flag of tabs: https://github.com/multani/sonata/issues/45
    * Fix hiding/showing the main window even if it's not the active widget:
      https://github.com/multani/sonata/issues/43
    * Deleting a track doesn't toggle the filter bar anymore in the current playlist.
    * Better support for MPD 0.18

Internal changes
''''''''''''''''

    * Simplified the `current` module. It should also now use less memory than before.
    * Provide a new GObject signal to notify on artwork change and artwork reset.


1.7a1 (2013-02-08)
------------------

New features
''''''''''''

    * Fetch covers from Last.fm instead of Rhapsody.com
    * Custom plugins are now stored along the configuration file in ~/.config/sonata/plugins/
    * Switch to Python 3 and Gtk 3 (Jörg Thalheim & Adam Dane)
    * More items in the tray menu (Kirill Lashuk)
    * Better fullscreen support (Kirill Lashuk)
    * Toggle fullscreen from the command line (Daniel)
    * Support MPD's "consume" mode (Anton Lashkov)
    * Use more default icons in the context menus (Anton Lashkov)
    * Load only the most recent of plugins with the same names

Bug fixes
'''''''''

    * Fix population of the "Save to playlist" menu (Zhihao Yuan)
    * Prevent multiple entries and improve art search with multi-CD albums (Kirill Lashuk)
    * Fixes weird show up if Sonata is not on the current workspace (Kirill Lashuk)
    * Scrobble after seeking to the beginning (Kirill Lashuk)
    * The lyricswiki plugin should now work again (Jonathan Ballet)
    * Fix UI status after reconnection to MPD (Jonathan Ballet)
    * Fix crash when searching the library (Kirill Lashuk)

Internal changes
''''''''''''''''

    * Simpler API for cover fetching plugins
    * Lot of code cleanup and internal changes
    * Removed Sugar UI support
    * Use Glade files to describe the interface and GtkBuilder to build the interface
    * More systematic configuration file management
    * High-level access to MPD's command results

Previous releases
-----------------

v1.6.2.1 - September 21, 2009
	+ Library artwork caching/displaying improvements
	+ Lyricwiki screen scraping workaround
	+ Replace Amazon.com cover art fetching with Rhapsody.com
	+ Bug: Fix library sorting when adding, e.g., albums to playlist
	+ Bug: Fix window positioning weirdness when unwithdrawing from systray
	+ Bug: Fix album being labeled various artists when a duplicate exists
	+ Bug: Fix album tracks in the info tab being out of order sometimes
	+ Bug: Show songs with a missing date in info album listing
	+ Bug: Library icons change after adding stream to playlist
	+ Bug: Cpu spike, tooltips missing in library search results with gtk 2.16.2

v1.6.2 - April 13, 2009
	+ Bug: Fix cli options (e.g. 'sonata play')

v1.6.1 - April 11, 2009
	+ Reduce library's filesystem view row height
	+ More auto-applying of preferences (e.g. format tab)
	+ Bug: Fix python version check for Ubuntu Jaunty
	+ Bug: Traceback if track has multiple date tags
	+ Bug: Fix infinite loop when format strings end in '{'
	+ Bug: Fix music directory chooser selecting the wrong folder sometimes

sonata.py
#!/usr/bin/env python
"""Sonata is a simple GTK+ client for the Music Player Daemon.
"""

__author__ = "Scott Horowitz"
__email__ = "stonecrest@gmail.com"
__license__ = """
Sonata, an elegant GTK+ client for the Music Player Daemon
Copyright 2006-2008 Scott Horowitz <stonecrest@gmail.com>

This file is part of Sonata.

Sonata is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 3 of the License, or
(at your option) any later version.

Sonata is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""

import sys, platform, locale, gettext, os

# XXX insert the correct sonata package dir in sys.path

try:
    import sonata
except ImportError:
    sys.stderr.write("Python failed to find the sonata modules.\n")
    sys.stderr.write("\nSearched in the following directories:\n" +
             "\n".join(sys.path) + "\n")
    sys.stderr.write("\nPerhaps Sonata is improperly installed?\n")
    sys.exit(1)

try:
    from sonata.version import version
except ImportError:
    sys.stderr.write("Python failed to find the sonata modules.\n")
    sys.stderr.write("\nAn old or incomplete installation was " +
             "found in the following directory:\n" +
             os.path.dirname(sonata.__file__) + "\n")
    sys.stderr.write("\nPerhaps you want to delete it?\n")
    sys.exit(1)

# XXX check that version.VERSION is what this script was installed for


## Apply global fixes:

# the following line is to fix python-zsi 2.0 and thus lyrics in ubuntu:
# https://bugs.launchpad.net/ubuntu/+source/zsi/+bug/208855
sys.path.append('/usr/lib/python2.5/site-packages/oldxml')

# hint for gnome.init to set the process name to 'sonata'
if platform.system() == 'Linux':
    sys.argv[0] = 'sonata'

# apply as well:
    try:
        import ctypes
        libc = ctypes.CDLL('libc.so.6')
        PR_SET_NAME = 15
        libc.prctl(PR_SET_NAME, sys.argv[0], 0, 0, 0)
    except Exception: # if it fails, it fails
        pass


## Apply locale and translation:

from sonata import misc
misc.setlocale()

# let gettext install _ as a built-in for all modules to see
# XXX what's the correct way to find the localization?
try:
    gettext.install('sonata', os.path.join(sonata.__file__.split('/lib')[0], 'share', 'locale'), unicode=1)
except:
    print "Warning: trying to use an old translation"
    gettext.install('sonata', '/usr/share/locale', unicode=1)
gettext.textdomain('sonata')


## Check initial dependencies:

# Test python version:
if sys.version_info < (2,5):
    sys.stderr.write("Sonata requires Python 2.5 or newer. Aborting...\n")
    sys.exit(1)

try:
    import mpd
except:
    sys.stderr.write("Sonata requires python-mpd. Aborting...\n")
    sys.exit(1)


## Initialize the plugin system:

from sonata.pluginsystem import pluginsystem
pluginsystem.find_plugins()
pluginsystem.notify_of('enablables',
               lambda plugin, cb: cb(True),
               lambda plugin, cb: cb(False))


## Load the command line interface:

from sonata import cli
args = cli.Args()
args.parse(sys.argv)

## Deal with GTK:

if not args.skip_gui:
    # importing gtk does sys.setdefaultencoding("utf-8"), sets locale etc.
    import gtk
    if gtk.pygtk_version < (2, 12, 0):
        sys.stderr.write("Sonata requires PyGTK 2.12.0 or newer. Aborting...\n")
        sys.exit(1)
    # fix locale
    misc.setlocale()
else:
    class FakeModule(object):
        pass
    # make sure the ui modules aren't imported
    for m in 'gtk', 'pango', 'sonata.ui', 'sonata.breadcrumbs':
        if m in sys.modules:
            print "Warning: module %s imported in CLI mode" % m
        else:
            sys.modules[m] = FakeModule()
    # like gtk, set utf-8 encoding of str objects
    reload(sys) # hack access to setdefaultencoding
    sys.setdefaultencoding("utf-8")


## Global init:

from socket import setdefaulttimeout as socketsettimeout
socketsettimeout(5)

if not args.skip_gui:
    gtk.gdk.threads_init()

    # we don't use gtk.LinkButton, but gtk.AboutDialog does;
    # in gtk 2.16.0 without this, the about uri opens doubly:
    gtk.link_button_set_uri_hook(lambda *args:None)

## CLI actions:

args.execute_cmds()


## Load the main application:

from sonata import main

app = main.Base(args)
try:
    app.main()
except KeyboardInterrupt:
    pass
